NuMPI/Optimization/polonsky_keer.py

```diff
@@ -0,0 +1,232 @@
+import numpy as np
+import scipy.optimize as optim
+import matplotlib as plt
+from inspect import signature
+
+
+def constrained_conjugate_gradients(objective, hessp, x0=None, gtol=1e-8,
+                                    mean_value=None, residual_plot=False,
+                                    maxiter=5000):
+    """
+    Implementation of constrained conjugate gradient algorithm as described in,
+    I.A. Polonsky, L.M. Keer, Wear 231, 206 (1999).
+
+    Parameters
+    ----------
+    objective : callable.
+                The objective function to be minimized.
+                            fun(x) -> float(energy),ndarray(gradient)
+                where x is the input ndarray.
+    hessp : callable
+            Function to evaluate the hessian product of the objective.
+            Hessp should accept either 1 argument (desscent direction) or
+            2 arguments (x,descent direction).
+                            hessp(des_dir)->ndarray
+                                    or
+                            hessp(x,des_dir)->ndarray
+            where x is the input ndarray and des_dir is the descent direction.
+
+    x0 : ndarray
+         Initial guess. Default value->None.
+         ValueError is raised if "None" is provided.
+
+    gtol : float, optional
```

Comment:
Write that the convergence criterion is max(abs) and norm2  of the projected gradient < gtol